pkg/logs/internal/decoder/auto_multiline_detection/tokenizer.go

```diff
@@ -0,0 +1,387 @@
+	tokens := make([]Token, 0, len(input))
+	if len(input) == 0 {
+		return tokens
+	}
+
+	run := 0
+	lastToken := getToken(input[0])
+	t.strBuf.Reset()
+	t.strBuf.WriteRune(unicode.ToUpper(rune(input[0])))
+
+	insertToken := func() {
+		defer func() {
+			run = 0
+			t.strBuf.Reset()
+		}()
+
+		// Only test for special tokens if the last token was a charcater (Special tokens are currently only A-Z).
+		if lastToken == c1 {
+			if t.strBuf.Len() == 1 {
+				if specialToken := getSpecialShortToken(t.strBuf.Bytes()[0]); specialToken != end {
+					tokens = append(tokens, specialToken)
+					return
+				}
+			} else if t.strBuf.Len() > 1 { // Only test special long tokens if buffer is > 1 token
+				if specialToken := getSpecialLongToken(t.strBuf.String()); specialToken != end {
+					tokens = append(tokens, specialToken)
+					return
+				}
+			}
+		}
+
+		// Check for char or digit runs
+		if lastToken == c1 || lastToken == d1 {
+			// Limit max run size
+			if run >= maxRun {
+				run = maxRun - 1
+			}
+			tokens = append(tokens, lastToken+Token(run))
+		} else {
+			tokens = append(tokens, lastToken)
+		}
+	}
+
+	for _, char := range input[1:] {
+		currentToken := getToken(char)
+		if currentToken != lastToken {
+			insertToken()
+		} else {
+			run++
+		}
+		if currentToken == c1 {
+			// Store upper case A-Z characters for matching special tokens
+			t.strBuf.WriteRune(unicode.ToUpper(rune(char)))
+		} else {
+			t.strBuf.WriteByte(char)
+		}
+		lastToken = currentToken
+	}
+
+	// Flush any remaining buffered tokens
+	insertToken()
+
+	return tokens
+}
+
+// getToken returns a single token from a single byte.
+func getToken(char byte) Token {
+	if unicode.IsDigit(rune(char)) {
+		return d1
+	} else if unicode.IsSpace(rune(char)) {
+		return space
+	}
+
+	switch char {
+	case ':':
+		return colon
+	case ';':
+		return semicolon
+	case '-':
+		return dash
+	case '_':
+		return underscore
+	case '/':
+		return fslash
+	case '\\':
+		return bslash
+	case '.':
+		return period
+	case ',':
+		return comma
+	case '\'':
+		return singlequote
+	case '"':
+		return doublequote
+	case '`':
+		return backtick
+	case '~':
+		return tilda
+	case '*':
+		return star
+	case '+':
+		return plus
+	case '=':
+		return equal
+	case '(':
+		return parenopen
+	case ')':
+		return parenclose
+	case '{':
+		return braceopen
+	case '}':
+		return braceclose
+	case '[':
+		return bracketopen
+	case ']':
+		return bracketclose
+	case '&':
+		return ampersand
+	case '!':
+		return exclamation
+	case '@':
+		return at
+	case '#':
+		return pound
+	case '$':
+		return dollar
+	case '%':
+		return percent
+	case '^':
+		return uparrow
+	}
+
+	return c1
+}
+
+func getSpecialShortToken(char byte) Token {
+	switch char {
+	case 'T':
+		return t
+	case 'Z':
+		return z
+	}
+	return end
+}
+
+// getSpecialLongToken returns a special token that is > 1 character
+func getSpecialLongToken(input string) Token {
+	switch input {
+	case "JAN", "FEB", "MAR", "APR", "MAY", "JUN", "JUL",
```

Comment:
That's a good question. And perhaps some context is missing with this PR standing on it's own. 
The tokenizer ultimately will serve 2 purposes. 

1. Assemble the token graph for timestamp detection. 
2. Compare any 2 log samples for similarity. 

In the case of `1` we only really need to support timestamp specific token types that occur in our list of "known" formats. Since this PR doesn't include that list - it's not obvious why this set of token types was chosen.

In the case of `2` - the datetime specific tokens matter a lot less. for example - if a log contains a a timezone not in the token list, it will fall back to a character run token. Since token string comparison will be approximate, it won't affect the behavior all that much. 

In short - special tokens make timestamp detection more accurate when paird with a set of "training" data (not in this PR). The tokenizer's usability for comparing logs is not generally negatively impacted by supporting more token types, and more token types will likely be added as we test on more real-world data. 

> Should all of this be configurable

IMO no - the tokenizer should have predictable behavior and it should not be changed without downstream tests in place to prevent accuracy regressions. In the next change I will be posting the token graph and set of "training" log data with unit tests to measure (and assert) the detection accuracy. By allowing someone to change the tokenizer behavior, it could cause a cascade of downstream effects in the detection logic that are very hard to debug. 

Instead - what will be configurable is: 
- the ability to disable timestamp detection if needed
- custom patterns from the user
- a match threshold for a custom pattern